tools/cli/plugin/context.go

```diff
@@ -0,0 +1,73 @@
+// The MIT License
+//
+// Copyright (c) 2021 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package plugin
+
+import (
+	"context"
+	"encoding/gob"
+
+	"google.golang.org/grpc/metadata"
+)
+
+var (
+	grpcIncomingMDKey = "grpc-incoming"
+	grpcOutgoingMDKey = "grpc-outgoing"
+)
+
+type PluginSafeContext struct {
+	Values map[string]interface{}
```

Comment:
How do you plan to use it? If it is just for metadata, why is it exported? If you plan to add some other values to `Values` why they are not extracted from `context.Context` and are not set back? This `PluginSafeContext` looks like a wrapper but it doesn't restore original context on `GetContext`.

Reply:
`Values` is exported so that `encoding/gob` will serialise/deserialize it, it ignores unexported fields. For now `Values` will hold just metadata, as that is the only context value that our current plugin APIs expect to be in the passed context. Later as we provide more plugin APIs we'd extend the list of values which we can send via PluginSafeContext.

Reply:
It may not be clear, but this reason I'm doing this at all is that `context.Context` is not serialisable by `encoding/gob`. There are some hacks like: https://github.com/raidancampbell/libraidan/blob/master/pkg/rruntime/context.go but that's pretty nasty (imho). So the alternative is that we transmit a known list of safely serialisable values via this mechanism and just add supported values as APIs require them or users request them.

Reply:
Thanks for clarifications, Rob.